### PR TITLE
RM-7807 Move AssemblyCompiler into its own project

### PR DESCRIPTION
**Pull request by KuJo7**
https://re-motion.atlassian.net/browse/RM-7807

### CODE REVIEW

File: Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ProjectType>Library</ProjectType>
+    <UseWindowsForms>true</UseWindowsForms>
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj:5]:
```suggestion
```

File: Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  </PropertyGroup>
+  <Import Project="..\..\..\Build\Shared.build.props" />
+  <PropertyGroup>
+    <AssemblyTitle>re-motion Development Compilation Library for Unit Tests</AssemblyTitle>
+    <Description>Library for unit testing, building, and other development-time activities.</Description>
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj:10]:
```suggestion
    <Description>Library for supporting compilation scenarios in unit tests.</Description>
```

File: Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    <AssemblyName>Remotion.Development.UnitTests.Compilation</AssemblyName>
+    <RootNamespace>$(AssemblyName)</RootNamespace>
+  </PropertyGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == 'net48'">
+  </ItemGroup>
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Development.UnitTests.Compilation.csproj:15]:
```suggestion
```

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// 
+using System;
+using System.Reflection;
+using System.Security;
+
+[assembly: AssemblyCulture("")]
+[assembly: CLSCompliant(true)]
+[assembly: AllowPartiallyTrustedCallers] // required for sandboxing
+
+[assembly: SecurityRules(SecurityRuleSet.Level1)]
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:25]: We probably don't need this
```suggestion

```

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Reflection;
+using System.Security;
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:19]:
```suggestion
```

File: Remotion/Globalization/UnitTests/Globalization.UnitTests.csproj [hunk 6]
```diff
@@ -18,12 +18,12 @@
...
     <ProjectReference Include="..\..\Core\Reflection\Core.Reflection.csproj" />
     <ProjectReference Include="..\..\Development\Core\Development.Core.csproj" />
+    <ProjectReference Include="..\..\Development\UnitTests.Compilation\Development.UnitTests.Compilation.csproj" />
     <ProjectReference Include="..\Core\Globalization.Core.csproj" />
   </ItemGroup>
 
   <Target Name="CopyTestAssemblyFilesToOutputPath" AfterTargets="Build" Condition="'$(TargetFramework)' != ''">
-    <Copy SourceFiles="$(MSBuildProjectDirectory)\$(IntermediateOutputPath)Remotion.Globalization.UnitTests.TestDomain.Resources.ClassWithResources.resources"
-          DestinationFiles="$(MSBuildProjectDirectory)\$(OutputPath)\TestDomain\TestAssemblies\En\En.Globalization.ClassWithResources.resources" />
+    <Copy SourceFiles="$(MSBuildProjectDirectory)\$(IntermediateOutputPath)Remotion.Globalization.UnitTests.TestDomain.Resources.ClassWithResources.resources" DestinationFiles="$(MSBuildProjectDirectory)\$(OutputPath)\TestDomain\TestAssemblies\En\En.Globalization.ClassWithResources.resources" />
```
Comment (a-ctor) [inline Remotion/Globalization/UnitTests/Globalization.UnitTests.csproj:26]: This is unrelated? Then we should reset it

**Review by a-ctor — CHANGES_REQUESTED** [review]

**Review by a-ctor — COMMENTED** [review]
Sadly, the name of the project was wrong in the issue description and has been changed to `Development.UnitTesting.Compilation`. `UnitTesting` is used instead of `UnitTests` to indicate that it contains infrastructure code that is relevant for unit testing but not unit tests. We need to fix this by changing the project names and namespaces.

File: Remotion/Development/UnitTesting.Compilation/UnitTesting/AssemblyCompilationException.cs [hunk 7]
```diff
@@ -17,7 +17,7 @@
 using System;
 using System.Runtime.Serialization;
 
-namespace Remotion.Development.UnitTesting
+namespace Remotion.Development.UnitTesting.Compilation.UnitTesting
```
Comment (a-ctor) [inline Remotion/Development/UnitTesting.Compilation/UnitTesting/AssemblyCompilationException.cs:20]: We should move this file and the one besides it out one level. We no longer need the `UnitTesting` part as suffix since we already have it before

File: Remotion/Development/UnitTesting.Compilation/UnitTesting/AssemblyCompiler.cs [hunk 8]
```diff
@@ -22,7 +22,7 @@
 using Microsoft.CSharp;
 using Remotion.Utilities;
 
-namespace Remotion.Development.UnitTesting
+namespace Remotion.Development.UnitTesting.Compilation.UnitTesting
 {
   [Serializable]
```
Comment (a-ctor) [inline Remotion/Development/UnitTesting.Compilation/UnitTesting/AssemblyCompiler.cs:27]: This is the second file that should move to the root.

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
```
Comment (a-ctor) [inline Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:10]: This file seems to be a leftover from the old project name.

**Review by a-ctor — CHANGES_REQUESTED** [review]

File: Remotion.sln [hunk 10]
```diff
@@ -351,6 +351,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Core.Development.Analyzers"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Core.Development.Analyzers.IntegrationTests", "Remotion\Core\Development.Analyzers.IntegrationTests\Core.Development.Analyzers.IntegrationTests.csproj", "{AD0BB354-A1E2-462C-BFCE-D6F0B32EC313}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Development.UnitTesting.Compilation", "Remotion\Development\UnitTesting.Compilation\Development.UnitTesting.Compilation.csproj", "{69723EC0-9306-446A-9CCA-E8F9114FC39B}"
```
Comment (MichaelKetting) [inline Remotion.sln:354]: Please add to projects.build file as release file.
cc @a-ctor

File: Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ProjectType>Library</ProjectType>
+  </PropertyGroup>
+  <Import Project="..\..\..\Build\Shared.build.props" />
+  <PropertyGroup>
+    <AssemblyTitle>re-motion Development Compilation Library for Unit Tests</AssemblyTitle>
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj:8]:
```suggestion
    <AssemblyTitle>re-motion Development Library for code compilation</AssemblyTitle>
```

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs (hunk 9 above)
Comment (MichaelKetting) [reply Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:10]: @a-ctor am confused. What did you find here in the license header?

File: Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    <Description>Library for supporting compilation scenarios in Unit Tests.</Description>
+    <AssemblyName>Remotion.Development.UnitTesting.Compilation</AssemblyName>
+    <RootNamespace>$(AssemblyName)</RootNamespace>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\Core\Core\Core.Core.csproj" />
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj:14]: Is the only reason for the core-dependency the argument checks? if so, we could (should now or later @a-ctor?) replace it with a shared import. That way, the library becomes more isolated, easier to use in stand alone scenarios

**Review by MichaelKetting — APPROVED** [review]
@a-ctor already found the major bits. I just had some message texts.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
.

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs (hunk 9 above)
Comment (a-ctor) [reply Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:10]: Hmm. Maybe it was displayed incorrectly I thought I saw double line comments `// // License ...` in one place.

File: Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj (hunk 12 above)
Comment (a-ctor) [reply Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj:14]: Yes only the argument checks are needed. I don't feel like it is necessary to remove the dependency on Core here. Chances are the consuming project already uses core if it uses the compilation library.

File: Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs (hunk 9 above)
Comment (MichaelKetting) [reply Remotion/Development/UnitTests.Compilation/Properties/AssemblyInfo.cs:10]: me too, just not here. at least in the code.

File: Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj (hunk 12 above)
Comment (MichaelKetting) [reply Remotion/Development/UnitTesting.Compilation/Development.UnitTesting.Compilation.csproj:14]: okay, so leave it for now. eventually, I will try to get most of the Development stuff Core free so the Core tests can use development without first needing to compile Core. It's just a bit of cyclic dependency that's annoying but not harmful at this point.

**Review by MichaelKetting — APPROVED** [review]

**Review by a-ctor — APPROVED** [review]
